### PR TITLE
CI: pyright in github actions

### PR DESCRIPTION
**Pull request by beauxq**
## What is this fixing or adding?

an alternative to https://github.com/ArchipelagoMW/Archipelago/pull/3107

pyright generally has fewer bugs than mypy

`pyright-config.json` has a list of files that will fail the CI if pyright finds errors in them

The idea is that we gradually add files and directories to the list as we get them cleaned up.
Eventually it can include all core python source files/directories.

World files and directories could be opt-in by the world maintainer. Adding all worlds doesn't need to be a goal.

This will help keep typing issues from creeping back in after they have been cleaned up.

The check that is done by CI is in a bash script `.github/type_check.sh`
- Developers can run this script before committing/pushing to see whether it will pass.
- Windows users should be able to run it with Git Bash.

A few minor typing changes are included here that are needed to finish cleaning these starting files.

## How was this tested?

running the `.github/type_check.sh` script

### CODE REVIEW

**Comment by black-sliver** [comment]
Couple of things off the bat:

First, I'm wondering if the pyright config should be top level, similar to a pytest.ini or .coveragerc, or a typical mypy.ini.
Also does it have to be json, or can it be toml? I'm wondering if it'd be worth switching to pyproject.toml at some point (note that pyproject can still just use setuptools and not touch our setup.py at all).

Second, and I assume the answer is no, would you happen to know if pyright can have a config per forlder, so worlds, tests or webhost can configure their runs of pyright separately, i.e. add or ignore files? The thing I have in mind would be like .gitignore, but I know that very few programs actually supports this kind of configuration.

Lastly, do you think a switch to pyright is preferred, and if so, should this replace #3107 or should we ideall merge both? Should we switch the "weak" check in our lint workflow to pyright as well?
I think having 2 optional strict checkers makes maintaining harder because it becomes less transparent, however I personally like mypy in general {for what i am doing elsewhere,if it wasn't for the issues we keep running into}.
One of the big maintenance issues I see is that if mypy follows some definitions that don't work, say world -> options, and the definition uses pyright because of that (options in pyright strict, not in mypy strict), the PR for options could still fail and the maintainer of options would have to also make world not fail.

Fwiw my main worries about pyright is that it doesn't use python, so it has a different platform compatibility, a different maintenance/security/update system, a different release schedule (compared to typings module) and a different incentives, however as long as vscod/ium stays relevant for python development, I expect pyright maintainers to put in the work.

**Comment by beauxq** [comment]
> First, I'm wondering if the pyright config should be top level, similar to a pytest.ini or .coveragerc, or a typical mypy.ini. Also does it have to be json, or can it be toml? I'm wondering if it'd be worth switching to pyproject.toml at some point (note that pyproject can still just use setuptools and not touch our setup.py at all).

If we have a `pyproject.toml` giving us a pyright configuration, I think that should check all files, for regular development work. But that would make a failure in CI for a long time, until we get everything cleaned up. That could be a long-term goal that we move towards.
This configuration that isn't the default would help us work towards that as we gradually add more files to it.
We could put this github actions config in the root of the repository if you want, but we should make sure it isn't picked up as the default, because we don't want to disable type checking on all the other files by default.

> Second, and I assume the answer is no, would you happen to know if pyright can have a config per forlder, so worlds, tests or webhost can configure their runs of pyright separately, i.e. add or ignore files? The thing I have in mind would be like .gitignore, but I know that very few programs actually supports this kind of configuration.

We could make a different json configuration, with a different pyright process for each one. I doubt a single pyright process will take configurations from multiple places.

> Lastly, do you think a switch to pyright is preferred, and if so, should this replace #3107 or should we ideally merge both? Should we switch the "weak" check in our lint workflow to pyright as well? I think having 2 optional strict checkers makes maintaining harder because it becomes less transparent, however I personally like mypy in general {for what i am doing elsewhere,if it wasn't for the issues we keep running into}. One of the big maintenance issues I see is that if mypy follows some definitions that don't work, say world -> options, and the definition uses pyright because of that (options in pyright strict, not in mypy strict), a PR for options could still fail and the maintainer of options would have to also make world not fail.

If we only use one, I think I have a slight preference towards pyright, because of the quality of the type checking.
I normally use both mypy and pyright for most of what I do.
It would be a little more complex, but I think it's feasible to put both checks in.
If we have one script that gives both mypy output and pyright output, it will be easier for developers to make sure they're satisfying both.
If we want to use both, I think this pyright PR should be merged first, and then I'll devise some script to read the files from the pyright config and feed them to mypy.

> Fwiw my main worries about pyright is that it doesn't use python, so it has a different platform compatibility, a different maintenance/security/update system, a different release schedule (compared to typings module) and different incentives, however as long as vscod/ium stays relevant for python development, I expect pyright maintainers to put in the work.

I didn't do this in either of these PRs, because I would like to see how bad the problem is before applying a solution that takes extra maintenance,
but I think we'll probably want to pin the versions of the type checkers to specific versions.
Neither mypy nor pyright is synchronized with `typing` module. The can both have new errors appear at any time.
I think we should see how much of a problem it is before we pin the version.

**Comment by black-sliver** [comment]
> [...] I think that should check all files, for regular development work.

Oh, somehow I was thinking that one would add files to the list before checking typing at all, but you are right that it's probably better to have "normal" pyright, and then the strict test completely separated.

> We could make a different json configuration, with a different pyright process for each one. I doubt a single pyright process will take configurations from multiple places.

Once we move worlds out of core, pulling the worlds + core in CI and being able to let the world decide what we should check, would be amazing to see when we broke something typing-wise.

> If we only use one, I think I have a slight preference towards pyright [...]

Alright, I'll spend some time with pyright then and review this PR (first) whenever I find the time.

**Comment by NewSoupVi** [comment]
> Second, and I assume the answer is no, would you happen to know if pyright can have a config per forlder, so worlds, tests or webhost can configure their runs of pyright separately

Ruff supports this, so it's worth checking whether pyright does

**Comment by black-sliver** [comment]
pyright made a change recently that makes it "unusable" for patterns we have.
I see 4 ways to move this forward:
1. wait for it to be resolved (or contribute ourselves to python.typing)
2. disable the check in question, which is not good because when we enable it again, we potentially have to remove a ton of files from the to-be-checked list
3. merge it and tell people not to use this feature yet
4. pin pyright to an older version

Also judging from the experience I just had, we should pin the version of pyright once we have a "good" one.

**Comment by beauxq** [comment]
For this `reportIncompatibleVariableOverride` issue, only 1 `# type: ignore` is required to deal with it for `worlds/zillion`.
(I have 2 `# type: ignore` to deal with it, but I realized only one is needed, because I don't need the other annotation for anything.)
I doubt any worlds will need more than 2 or 3.
I think this an ok workaround for the time being, and I wouldn't consider it unusable.

**Comment by black-sliver** [comment]
We use the same pattern in classes inheriting from WorldTestBase, for example. So while it is true that impact is quite small in some cases, it is both annoying and possibly hiding problems in others. The ways to work around it (say, not typing it and then `assert(isinstance(...))`) are kinda defeating the advantages of the type hints for me, because you add boiler plate and then still not express the thing you want to (i.e. `world: SoEWorld` in class body).

I think pinning pyright to a version before #5955 was merged would be better. So v1.1.327, I think.

This is happening is also a good example why we should pin it in general. Pyright changed behavior without using semantic versioning. We can argue that it was wrong before, however the only way to express our typing is not supported anymore, even if it was technically never precise enough.

I'd rather have the PR "bump pyright" fail than all PRs after an automatic pyright update and I'd rather have an imprecise type annotation we fix later than adding type: ignore that will stay forever.

**Comment by black-sliver** [comment]
Note that what I am saying here doesn't stop you from adding the type: ignore, keep up to date with latest typing conventions and using latest pyright locally, but that's different from CI, where a breaking change will lead to pyright getting kicked out again because that's easier than working out what needs to change.

**Comment by black-sliver** [comment]
Oh, I should've said "**other** work-arounds". I think type: ignore is the proper solution here if you want to use latest pyright.

**Comment by beauxq** [comment]
So is this usage of `# type: ignore` bad enough that we should pin pyright to an older version, losing features like [improved metaclass support](https://github.com/microsoft/pyright/pull/7387) among other fixes and features.

(I already said I think we should pin the version, but the concern is whether it's worth it to pin to this older version.)

**Comment by black-sliver** [comment]
Is this a problem we run into currently? Picking which version  to pin is still to be decided, however I would prefer to pin something that doesn't have said change.

**Comment by beauxq** [comment]
> Is this a problem we run into currently?

I'm not sure what this is referring to. That metaclass improvement does affect things we have in Archipelago. That's why I reported the issue to pyright, and they made the improvement.

I don't think it's worth it to lose all the later fixes and features pinning a version that doesn't have the override change, just so that we don't have to put `# type: ignore` on those overrides.

Also, keep in mind that the main focus of this endeavor is for core typing - with the side-effect that worlds can opt in.
I have yet to see that override issue in core code. (I acknowledge it's probably there somewhere, but much less prominent.) That makes it even less worth it to pin a version that doesn't have that change.

File: .github/workflows/strict-type-check.yml
Comment (black-sliver) [inline .github/workflows/strict-type-check.yml:9]: this should also trigger for `.github/type_check.sh`, `.github/pyright-config.json`, `.github/workflows/strict-type-check.yml` and on `**.pyi`

File: .github/type_check.sh
Comment (black-sliver) [inline .github/type_check.sh:5]: Do you feel confident in porting this to .bat for our windows users? While awful from a maintenance standpoint, I know of at least one regular contributor that does not have any bash installed. You can test the .bat without running Windows by switching `runs-on` to `windows-latest` in the gh action.

Comment (beauxq) [reply .github/type_check.sh]: How about this change.
Now it uses python instead of bash or bat.
I think that will be better if we add mypy anyway.

**Comment by beauxq** [comment]
I'm going to push an error to test and make sure it fails.

**Review by black-sliver — APPROVED** [review]
Thanks!

**Pull request merged**
